trust-quorum/src/proxy.rs

```diff
@@ -0,0 +1,589 @@
+            Ok(val) => Ok(val),
+        };
+
+        f(res)
+    }
+}
+
+/// The error result of a proxy request
+#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq, SlogInlineError)]
+pub enum ProxyError<T: std::error::Error> {
+    #[error(transparent)]
+    Inner(#[from] T),
+    #[error("disconnected")]
+    Disconnected,
+    #[error("response for different type received: {0}")]
+    InvalidResponse(String),
+    #[error("task sender dropped")]
+    RecvError,
+    #[error("task is busy: channel full")]
+    Busy,
+}
+
+/// An error returned from a [`Tracker`].
+///
+/// This error wraps a `WireError` so that it can also indicate when no response
+/// was received, such as when the sprockets channel was disconnected.
+pub enum TrackerError {
+    Wire(WireError),
+    Disconnected,
+}
+
+/// A trackable in-flight proxy request, owned by the `Tracker`
+#[derive(Debug)]
+pub struct TrackableRequest {
+    request_id: Uuid,
+    destination: BaseboardId,
+    // The option exists so we can take the sender out in `on_disconnect`, when
+    // the request is borrowed, but about to be discarded.
+    tx: DebugIgnore<Option<oneshot::Sender<Result<WireValue, TrackerError>>>>,
+}
+
+impl TrackableRequest {
+    pub fn new(
+        destination: BaseboardId,
+        request_id: Uuid,
+        tx: oneshot::Sender<Result<WireValue, TrackerError>>,
+    ) -> TrackableRequest {
+        TrackableRequest { request_id, destination, tx: DebugIgnore(Some(tx)) }
+    }
+}
+
+impl IdHashItem for TrackableRequest {
+    type Key<'a> = &'a Uuid;
+
+    fn key(&self) -> Self::Key<'_> {
+        &self.request_id
+    }
+
+    id_upcast!();
+}
+
+/// A mechanism to keep track of proxied requests and wait for their replies
+pub struct Tracker {
+    // In flight operations
+    ops: IdHashMap<TrackableRequest>,
+}
+
+impl Tracker {
+    pub fn new() -> Tracker {
+        Tracker { ops: IdHashMap::new() }
+    }
+
+    /// The number of proxied requests outstanding
+    pub fn len(&self) -> usize {
+        self.ops.len()
+    }
+
+    pub fn insert(&mut self, req: TrackableRequest) {
+        self.ops.insert_unique(req).expect("no duplicate request IDs");
+    }
+
+    /// Handle a proxied response from a peer
+    pub fn on_response(
+        &mut self,
+        request_id: Uuid,
+        result: Result<WireValue, WireError>,
+    ) {
+        if let Some(mut req) = self.ops.remove(&request_id) {
+            let res = result.map_err(TrackerError::Wire);
+            let _ = req.tx.take().unwrap().send(res);
+        }
+    }
+
+    /// A remote peer has disconnected
+    pub fn on_disconnect(&mut self, from: &BaseboardId) {
+        self.ops.retain(|mut req| {
+            if &req.destination == from {
+                let tx = req.tx.take().unwrap();
+                let _ = tx.send(Err(TrackerError::Disconnected));
+                false
+            } else {
+                true
+            }
+        });
```

Comment:
ah, this is a place where it would be nice to have a bijective multimap so we could do this operation without iterating over every in-flight op, isn't it...?

but, this makes sense in the absence of that.